rust/integration-tests/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+            .iter()
+            .enumerate()
+            .map(|(idx, txn)| {
+                let mut txn: Transaction = serde_json::from_slice(txn).unwrap();
+                txn.version = idx as u64;
+                txn
+            })
+            .collect::<Vec<Transaction>>();
+        let postgres_container = GenericImage::new("postgres", "14")
+            .with_exposed_port(5432.tcp())
+            .with_wait_for(WaitFor::message_on_stderr(
+                "database system is ready to accept connections",
+            ))
+            .with_env_var("POSTGRES_DB", "postgres")
+            .with_env_var("POSTGRES_USER", "postgres")
+            .with_env_var("POSTGRES_PASSWORD", "postgres")
+            .start()
+            .await
+            .expect("Redis started");
+        Ok(TestContext {
+            transaction_batches,
+            postgres_container,
+        })
+    }
+
+    async fn create_schema(&self) -> anyhow::Result<()> {
+        let db_url = self.get_db_url().await;
+        let mut conn = PgConnection::establish(&db_url)
+            .with_context(|| format!("Error connecting to {}", db_url))?;
+        // Drop the schema and recreate it.
+        sql_query("DROP SCHEMA public CASCADE;")
+            .execute(&mut conn)
+            .unwrap();
+        sql_query("CREATE SCHEMA public;")
+            .execute(&mut conn)
+            .unwrap();
+        run_pending_migrations(&mut conn);
+        Ok(())
+    }
+
+    pub async fn get_db_url(&self) -> String {
+        let host = self.postgres_container.get_host().await.unwrap();
+        let port = self
+            .postgres_container
+            .get_host_port_ipv4(5432)
+            .await
+            .unwrap();
+        format!("postgres://postgres:postgres@{host}:{port}/postgres")
+    }
+
+    // `run` functions takes a closure that is executed after the test context is created.
+    // The closure will be executed multiple times with different permutations of the transactions.
+    // For example:
+    //   test.run(async move | context | {
+    //       // Runs after every permutatation
+    //       let res = Diesel::raw_sql(conn, "select amount from balances where user = '0x1'");
+    //       assert_eq!(res["amount"], 100, "winner balance incorrect when txn order: {:?}", context.txn_order);
+    //   }).await;
+    //
+    pub async fn run<F>(
+        &self,
+        processor_config: TestProcessorConfig,
+        verification_f: F,
+    ) -> anyhow::Result<()>
+    where
+        F: Fn(&mut PgConnection) -> anyhow::Result<()> + Send + Sync + 'static,
+    {
+        let transactions = self.transaction_batches.clone();
+        let db_url = self.get_db_url().await;
+        let mut conn = PgConnection::establish(&db_url)
+            .with_context(|| format!("Error connecting to {}", db_url))?;
+        let db_pool = new_db_pool(&db_url, None).await.unwrap();
+
+        for perm in transactions.iter().permutations(transactions.len()) {
```

Comment:
Could we have a comment here explaining why we're doing the permutations, ie to test that transactions processed out of order will result in the same output in DB.